esrally/track/loader.py

```diff
@@ -391,55 +395,40 @@ def used_corpora(t):
     return corpora.values()
 
 
-def prepare_track(t, cfg):
-    """
-    Ensures that all track data are available for running the benchmark.
-
-    :param t: A track that is about to be run.
-    :param cfg: The config object.
-    """
-    data_root_dir = cfg.opts("benchmarks", "local.dataset.cache")
-    tp = CompositeTrackProcessor(cfg)
-    logger = logging.getLogger(__name__)
-
-    logger.info("Preparing track [%s]", t.name)
-    if t.has_plugins:
-        logger.info("Reloading track [%s] to ensure plugins are up-to-date.", t.name)
-        # the track might have been loaded on a different machine (the coordinator machine) so we force a track update
-        # to ensure we use the latest version of plugins.
-        load_track_plugins(cfg, t.name, register_track_processor=tp.register_track_processor, force_update=True)
-
-    # register last so user-defined track processors can override the default behavior
-    tp.register_track_processor(DefaultTrackPreparator(cfg))
-
-    tp.on_prepare_track(t, data_root_dir)
-
-
 class DefaultTrackPreparator(TrackProcessor):
     def __init__(self, cfg):
         super().__init__()
         self.cfg = cfg
-        self.logger = logging.getLogger(__name__)
         # just declare here, will be injected later
         self.downloader = None
         self.decompressor = None
+        self.track = None
+
+    @staticmethod
+    def prepare_docs(cfg, track, corpus, preparator):
+        for document_set in corpus.documents:
+            if document_set.is_bulk:
+                data_root = data_dir(cfg, track.name, corpus.name)
+                logging.getLogger(__name__).info("Resolved data root directory for document corpus [%s] in track [%s] "
+                                                 "to [%s].", corpus.name, track.name, data_root)
+                if len(data_root) == 1:
+                    preparator.prepare_document_set(document_set, data_root[0])
+                # attempt to prepare everything in the current directory and fallback to the corpus directory
+                elif not preparator.prepare_bundled_document_set(document_set, data_root[0]):
+                    preparator.prepare_document_set(document_set, data_root[1])
 
     def on_prepare_track(self, track, data_root_dir):
+        self.track = track
```

Comment:
Is a field necessary here? As far as I could see `track` is only used locally?

Reply:
I was debugging this line, `track` is poisoned because of `from esrally.track import track`.  I'll make this clean without setting an instance field